presto-native-execution/presto_cpp/main/PrestoServer.cpp

```diff
@@ -1401,6 +1410,45 @@ static protocol::Duration getUptime(
   return protocol::Duration(seconds, protocol::TimeUnit::SECONDS);
 }
 
+void PrestoServer::convertToVeloxPlan(
+    proxygen::HTTPMessage* message,
+    proxygen::ResponseHandler* downstream,
+    const std::vector<std::unique_ptr<folly::IOBuf>>& body) {
+  std::string error;
+  try {
+    auto headers = message->getHeaders();
+
+    std::ostringstream oss;
+    for (auto& buf : body) {
+      oss << std::string((const char*)buf->data(), buf->length());
+    }
+    std::string planFragmentJson = oss.str();
+    protocol::PlanFragment planFragment = json::parse(planFragmentJson);
+
+    auto queryCtx = core::QueryCtx::create();
+    VeloxInteractiveQueryPlanConverter converter(queryCtx.get(), pool_.get());
+
+    // Create static taskId and empty TableWriteInfo needed for plan conversion
+    protocol::TaskId taskId = "velox-plan-conversion.0.0.0";
+    auto tableWriteInfo = std::make_shared<protocol::TableWriteInfo>();
+
+    // Attempt to convert the plan fragment to a Velox plan
+    converter.toVeloxQueryPlan(planFragment, tableWriteInfo, taskId);
+
+  } catch (VeloxException& e) {
+    error = e.message();
+  } catch (std::exception& e) {
+    error = e.what();
+  }
+
+  // Return ok status if conversion succeeded or error if failed
+  if (error.empty()) {
+    http::sendOkResponse(downstream, json(R"({ "status": "ok" })"));
+  } else {
+    http::sendErrorResponse(downstream, json(R"({ "status": "error", "message": ")" + error + R"(")})"));
```

Comment:
We can make the response however you think is best. I didn't put too much thought into the message since we discussed that this would eventually be sending the converted plan back. Are you saying if the plan validation fails you would send an OkResponse and not an ErrorResponse?

Reply:
yeah, i think we should follow the model for how errors get passed along in general for native workers, which i think is an OkResponse with the error information in the TaskInfo.

Reply:
@rschlussel if I understand correctly, you are asking for the response to be a `TaskInfo` message with `ExecutionFailureInfo` to be populated if the conversion fails?  

On the Java side, `TaskInfo` and many of the related classes are not part of the SPI, so things will need to be moved around. Is that what we want here?

On the C++ side, there is lots of information in the `TaskInfo` struct, I'm not sure how much is really useful in this case and if it needs to be populated with the `TaskManager` and have a corresponding `PrestoTask`.

Would it make sense to define a simple response message? cc @tdcmeehan

Reply:
To me, the difference with tasks is that 4xx is returned when we can not create the task.  What is being reported by the TaskInfo is whether or not the already successfully created task ended up failing.  That should be returned as a 200, because the call to check on the task status itself succeeded, and it is merely reporting the underlying failure.  When we go to update or create the task, when 4xx is returned, we are returning that only when the task cannot be created.

In this case, what is being returned is the converted plan.  That should probably return [code 422](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422), because we are returning whether or not the conversion failed.  Like all 4xx error codes, this indicates the problem is with the input.  It's similar to a task update or creation failing.